orchestra/utils/view_helpers.py

```diff
@@ -12,3 +14,93 @@ class IsAssociatedWorker(permissions.BasePermission):
     def has_object_permission(self, request, view, obj):
         worker = Worker.objects.get(user=request.user)
         return obj.worker == worker
+
+
+def _get_changed_fields(old_todo, new_todo):
+    changed_fields = []
+    if old_todo.title != new_todo.title:
+        changed_fields.append('title')
+    if old_todo.details != new_todo.details:
+        changed_fields.append('details')
+    dict1 = {}
+    dict2 = {}
+    if isinstance(old_todo.additional_data, dict):
+        dict1 = old_todo.additional_data
+    if isinstance(new_todo.additional_data, dict):
+        dict2 = new_todo.additional_data
+    changed_subfields = [k for k in dict1.keys()
+                         if dict1.get(k) != dict2.get(k)]
+    changed_fields.extend(changed_subfields)
+    return changed_fields
+
+
+def get_todo_change(old_todo, new_todo):
+    # When activity_log is updated, `todo_change = None`
+    # to avoid triggering any slack messages
+    todo_change = ''
+    changed_fields = _get_changed_fields(old_todo, new_todo)
+    if old_todo.completed != new_todo.completed:
+        todo_change = 'complete' if new_todo.completed else 'incomplete'
+    elif old_todo.skipped_datetime != new_todo.skipped_datetime:
+        todo_change = 'not relevant' \
+            if new_todo.skipped_datetime else 'relevant'
+    message = ''
+    if len(todo_change) and len(changed_fields):
+        message = '{}. Changed fields: {}'.format(
+            todo_change, ', '.join(changed_fields))
+    elif len(changed_fields):
+        message = 'changed. Changed fields: {}'.format(
+            ', '.join(changed_fields))
+    else:
+        message = todo_change
+    return message
+
+
+def _get_sender(user):
+    if isinstance(user, SignedUser):
+        return None
+    return Worker.objects.get(
+        user=user).formatted_slack_username()
+
+
+def notify_single_todo_update(user, old_todo, todo):
+    # To avoid Slack noise, only send updates for changed TODOs with
+    # depth 0 (no parent) or 1 (no grandparent).
+    todo_change = get_todo_change(old_todo, todo)
+    sender = _get_sender(user)
+    if todo_change and \
+            (not (todo.parent_todo and todo.parent_todo.parent_todo)):
+        if sender:
+            message = '{} has marked `{}` as `{}`.'.format(
```

Comment:
making sure i get how all the messages come together:

"`X` has marked `some todo` as `not relevant. Changed Fields: title.`."

I suspect we want something closer to
"`X` has updated `some todo`: marked `[complete|incomplete|relevant|not relevant]`, changed `[title|description|title & description]`."

Reply:
Addressed. Wrote some tests. Hopefully, the new messages sound natural to native ears :)